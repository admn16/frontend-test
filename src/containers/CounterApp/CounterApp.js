import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { hot } from 'react-hot-loader';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import styled from 'styled-components';
import CounterForm from 'components/CounterForm/CounterForm';
import Counters from 'components/Counters/Counters';
import Total from 'components/Total/Total';
import * as counterActions from 'actions/counterActions';
import * as uiActions from 'actions/uiActions';

const StyledApp = styled.main`
  margin: 0 auto;
  padding-top: 20px;
  width: 400px;
`;

class CounterApp extends PureComponent {
  static propTypes = {
    getCounters: PropTypes.func.isRequired,
    incrementCounter: PropTypes.func.isRequired,
    decrementCounter: PropTypes.func.isRequired,
    deleteCounter: PropTypes.func.isRequired,
    addCounter: PropTypes.func.isRequired,
    updateText: PropTypes.func.isRequired,

    counters: PropTypes.arrayOf(PropTypes.object),
    text: PropTypes.string.isRequired,
  };

  static defaultProps = {
    counters: [],
  };

  componentDidMount() {
    const { getCounters } = this.props;
    getCounters();
  }

  onAdd = (e) => {
    e.preventDefault();

    const { addCounter, text } = this.props;

    if (text) {
      addCounter(text);
    }
  }

  onDelete = id => () => {
    const { deleteCounter } = this.props;
    deleteCounter(id);
  };

  onDecrement = id => () => {
    const { decrementCounter } = this.props;
    decrementCounter(id);
  };

  onIncrement = id => () => {
    const { incrementCounter } = this.props;
    incrementCounter(id);
  };

  onUpdateText = ({ target }) => {
    const { updateText } = this.props;
    updateText(target.value);
  }

  render() {
    const { counters, text } = this.props;
    const total = counters.length
      ? counters
        .map(i => i.count)
        .reduce((acc, curr) => acc + curr)
      : 0;

    return (
      <StyledApp>
        <CounterForm
          onChange={this.onUpdateText}
          onSubmit={this.onAdd}
          text={text}
        />
        <Counters
          counters={counters}
          onDelete={this.onDelete}
          onDecrement={this.onDecrement}
          onIncrement={this.onIncrement}
        />
        <Total total={total} />
      </StyledApp>
    );
  }
}

const mapStateToProps = state => ({
  counters: state.counters,
  text: state.ui.text,
});

const mapDispatchToProps = dispatch => bindActionCreators({
  getCounters: counterActions.fetchCounters,
  incrementCounter: counterActions.incrementCounter,
  decrementCounter: counterActions.decrementCounter,
  deleteCounter: counterActions.deleteCounter,
  addCounter: counterActions.addCounter,
  updateText: uiActions.updateText,
}, dispatch);

export { CounterApp };

export default hot(module)(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(CounterApp),
);
