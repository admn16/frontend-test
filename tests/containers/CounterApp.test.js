import React from 'react';
import renderer from 'react-test-renderer';
import { CounterApp } from 'containers/CounterApp/CounterApp';

describe('<CounterApp/>', () => {
  const mockFn = jest.fn();
  const props = {
    getCounters: mockFn,
    incrementCounter: mockFn,
    decrementCounter: mockFn,
    deleteCounter: mockFn,
    addCounter: mockFn,
    updateText: mockFn,

    counters: [{ count: 1, title: 'test123', id: '123asd' }],
    text: 'test 123123',
  };

  test('renders correctly', () => {
    const tree = renderer
      .create(<CounterApp {...props} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
