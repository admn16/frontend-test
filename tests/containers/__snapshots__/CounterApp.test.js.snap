// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CounterApp/> renders correctly 1`] = `
<main
  className="CounterApp__StyledApp-c6b218-0 bnetwR"
>
  <div
    className="Card__StyledCard-hu4gdg-0 hNyZiT"
  >
    <form
      className="CounterForm__Form-sc-1w9o0z7-0 emKkNx"
      onSubmit={[Function]}
    >
      <input
        className="CounterForm__Input-sc-1w9o0z7-1 fHQEKr"
        onChange={[Function]}
        placeholder="Add new counter"
        value="test 123123"
      />
      <button
        className="CounterForm__Button-sc-1w9o0z7-2 iXLQka"
        type="submit"
      >
        +
      </button>
    </form>
  </div>
  <div
    className="Counters__StyledCounters-vlq2we-0 iBWZCO Card__StyledCard-hu4gdg-0 hNyZiT"
  >
    <article>
      <div
        className="Counter__StyledCounter-sc-1q55mxl-0 eWqOki"
      >
        <button
          className="Counter__Delete-sc-1q55mxl-3 foqxoZ"
          onClick={[Function]}
          type="button"
        >
          <i
            className="fas fa-trash-alt"
          />
        </button>
        <span
          className="Counter__Title-sc-1q55mxl-1 gzIbEF"
        >
          test123
        </span>
        <button
          className="Counter__Increment-sc-1q55mxl-4 gSvDCI Counter__Delete-sc-1q55mxl-3 foqxoZ"
          onClick={[Function]}
          type="button"
        >
          <i
            className="fas fa-plus"
          />
        </button>
        <span
          className="Counter__Count-sc-1q55mxl-2 egAVVf"
        >
          1
        </span>
        <button
          className="Counter__Decrement-sc-1q55mxl-5 cJKjoF Counter__Increment-sc-1q55mxl-4 gSvDCI Counter__Delete-sc-1q55mxl-3 foqxoZ"
          onClick={[Function]}
          type="button"
        >
          <i
            className="fas fa-minus"
          />
        </button>
      </div>
    </article>
  </div>
  <div
    className="Total__StyledTotal-i597gn-0 glAgzs Card__StyledCard-hu4gdg-0 hNyZiT"
  >
    <p
      className="Total__Text-i597gn-1 gYVLGj"
    >
      Total Sum:Â 
      <span
        className="Total__TotalText-i597gn-2 hhdlqZ"
      >
        1
      </span>
    </p>
  </div>
</main>
`;
